typed/symdiff_test.go

```diff
@@ -0,0 +1,483 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package typed
+
+import (
+	"fmt"
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/schema"
+	"sigs.k8s.io/structured-merge-diff/value"
+
+	"gopkg.in/yaml.v2"
+)
+
+type symdiffTestCase struct {
+	name         string
+	rootTypeName string
+	schema       string
+	quads        []symdiffQuad
+}
+
+type symdiffQuad struct {
+	lhs      string
+	rhs      string
+	modified *fieldpath.Set
+	added    *fieldpath.Set
+}
+
+var symdiffCases = []symdiffTestCase{{
+	name:         "simple pair",
+	rootTypeName: "stringPair",
+	schema: `types:
+- name: stringPair
+  struct:
+    fields:
+    - name: key
+      type:
+        scalar: string
+    - name: value
+      type:
+        untyped: {}
+`,
+	quads: []symdiffQuad{{
+		`{"key":"foo","value":1}`,
+		`{"key":"foo","value":1}`,
+		_NS(),
+		_NS(),
+	}, {
+		`{"key":"foo","value":{}}`,
```

Comment:
Since value is "untyped"/atomic, maybe add a new test that shows that sub-fields are not tracked specifically in the diff

Reply:
test added